electrumsv/cached_headers.py

```diff
@@ -1,82 +1,260 @@
+    for header_index in range(last_index + 1, len(headers)):
+        read_header(header_index)
+
+
+
+# NOTE(typing) `bitcoinx.Headers` is untyped. We have to ignore the typing error telling us that.
+class ElectrumSVHeaders(Headers): # type: ignore[misc]
+    """
+    We only use this class if we are reconstituting the metadata from a matching data file.
+    """
+    def __init__(self, network: Network, file_path: str, checkpoint: CheckPoint) -> None:
+        """
+        The core reason for using this is to bypass the initial processing of the headers file
+        so that we can use the chain data instead.
+        """
+        self.common_setup(network, file_path, checkpoint)
+
+        self._chains: list[Chain] = []
+        self._short_hashes = bytearray()
+        self._heights = array.array('I')
+        self._chain_indices = array.array('H')
+
+
+def hash_headerfile(file_path: str, file_length: int) -> bytes:
     sha256_hash = sha256()
-    hf = open(file_path, "rb")
-    try:
-        while True:
-            headers_chunk = hf.read(65536)
+    with open(file_path, "rb") as hf:
+        reserved_bytes = hf.read(bitcoinx.chain._HeaderStorage.struct_reserved.size)
+        actual_reserved_size, header_store_version, header_store_count = \
+            bitcoinx.chain._HeaderStorage.struct_reserved.unpack(reserved_bytes)
+        # NOTE(rt12) If this starts erroring you need to consider what it means for our
+        #     incremental chain data. bitcoinx would hopefully migrate.
+        assert header_store_version == 0
+        file_length -= bitcoinx.chain._HeaderStorage.struct_reserved.size
+
+        while file_length > 0:
+            headers_chunk = hf.read(min(65536, file_length))
             if not headers_chunk:
+                assert file_length == 0
                 break
             sha256_hash.update(headers_chunk)
-    finally:
-        hf.close()
+            file_length -= len(headers_chunk)
     return sha256_hash.digest()
 
 
+class CachedHeadersMetadata(NamedTuple):
+    version: int
+    headerfile_length: int
+    headerfile_hash: bytes
+
+class CachedHeadersChainData(NamedTuple):
+    parent_first_height: int
+    first_height: int
+    tip_height: int
+    tip_header_bytes: bytes
+    work: int
+    header_indices: array.array[int]
+
+class CachedHeadersData(NamedTuple):
+    last_index: int
+    short_hashes: bytearray
+    heights: array.array[int]
+    chain_indices: array.array[int]
+
+
+def read_cached_headers_metadata(f: io.BufferedReader) -> CachedHeadersMetadata:
+    chaindata_version = read_le_uint16(f.read)
+    assert chaindata_version == 1
+    headerfile_length = read_le_uint32(f.read)
+    headerfile_hash = f.read(32)
+    assert len(headerfile_hash) == 32
+    return CachedHeadersMetadata(chaindata_version, headerfile_length, headerfile_hash)
+
+def write_cached_headers_metadata(f: io.BufferedWriter, file_length: int, file_hash: bytes) -> None:
+    f.write(pack_le_uint16(1))
```

Comment:
I think it would be worth using a constant variable here like `HEADERS_STORE_VERSION=1`

Reply:
We can use the other comment to cover this.